packages/data-controls/src/index.js

```diff
@@ -91,19 +107,134 @@ export function dispatch( ...args ) {
  * import * as resolvers from './resolvers';
  *
  * registerStore( 'my-custom-store', {
- * 	reducer,
- * 	controls,
- * 	actions,
- * 	selectors,
- * 	resolvers,
+ * reducer,
+ * controls,
+ * actions,
+ * selectors,
+ * resolvers,
  * } );
  * ```
- *
  * @return {Object} An object for registering the default controls with the
- *                  store.
+ * store.
  */
 export const controls = {
+	ACQUIRE_STORE_LOCK: ( { scope, exclusive } ) => {
+		const promise = new Promise( ( resolve ) => {
+			lockRequests.unshift( {
+				exclusive,
+				scope,
+				notifyAcquired: resolve,
+			} );
+		} );
+		acquireAvailableLocks();
+		return promise;
+	},
+
+	RELEASE_STORE_LOCK: ( { lock } ) => {
+		releaseLock( lock );
+		// Run anything that got unblocked by releasing this lock in the next tick:
+		setTimeout( () => {
+			acquireAvailableLocks();
+		} );
+	},
+
 	API_FETCH( { request } ) {
 		return triggerFetch( request );
 	},
 };
+
+const lockRequests = [];
+const lockTree = {
+	locks: [],
+	children: {},
+};
+
+function acquireAvailableLocks() {
+	for ( let i = lockRequests.length - 1; i >= 0; i-- ) {
```

Comment:
Why is this queue "backwards"? Usually when I see a queue in JavaScript it's `push` for enqueue and `shift` for dequeue.

Reply:
This was just a convenient way of removing items from a list while iterating through it